### ISSUE TITLE
Is it possible to eliminate Compose and Identity like in Idris?

### ISSUE THREAD

**Issue by 2426021684** [VinylRecords/Vinyl#118]
Compare this haskell code to the following idris code. Is there a way to eliminate the `Compose` and `Identity` from the haskell code? The idris can not be directly translated to haskell due to the lack of partially applied type families.

```haskell
module Main where

import Data.Vinyl
import Data.Maybe
import Data.Vinyl.Functor

myFunction :: Rec (Maybe :. f) ts -> Rec f ts -> Rec f ts
myFunction = rzipWith (flip fromMaybe . getCompose)

test :: Rec Identity [String, Int]
test = myFunction (Compose (Just . Identity $ "hi") :& Compose Nothing :& RNil) (Identity "bye" :& Identity 4 :& RNil)
```

```idris
module Main

import Data.Vect
import Data.Vect.Quantifiers

Id : Type -> Type
Id = id

rzipWith : ({t : _} -> a t -> b t -> c t) -> All a ts -> All b ts -> All c ts
rzipWith f [] [] = []
rzipWith f (a::as) (b::bs) = f a b :: rzipWith f as bs

myFunction : All (Maybe . f) ts -> All (Lazy . f) ts -> All f ts
myFunction = rzipWith (flip fromMaybe)

test : All Id [String, Int]
test = myFunction [Just "hi", Nothing] ["bye", 4]
```

**Comment by acowley** [comment]
We should look into whether we can take advantage of the recently-discussed [HKD](http://reasonablypolymorphic.com/blog/higher-kinded-data) technique.

**Comment by 2426021684** [comment]
@acowley Interesting but that does not handle `Compose` although it is possible that it could be extended too. It might be a good idea to handle `Const` too.

**Comment by 2426021684** [comment]
@acowley 

What do you think of this? Since `RunX` is an open type family you can extend it with what ever you need. It might be possible to get `:&` to apply `XData` for you. `XMagic` needs a better name.

```haskell
{-# LANGUAGE OverloadedStrings, RecursiveDo, LambdaCase, GADTs, DataKinds,
   TypeOperators, KindSignatures, TypeFamilies, RankNTypes, TypeApplications,
   ScopedTypeVariables, MultiParamTypeClasses, FlexibleContexts,
   FlexibleInstances, UndecidableInstances, PartialTypeSignatures, PolyKinds
   #-}
module Main where

import Data.Maybe
import GHC.TypeLits
import Data.Vinyl
import Data.Vinyl.Functor

data XId :: (k -> k -> *)
data XConst (a :: k1) :: (k -> k1 -> *)
data XMagic (a :: k1 -> (k1 -> k2) -> *) :: k1 -> k2 -> *
data XCompose (a :: k1 -> k2 -> *) (b :: k0 -> k1 -> *) :: k0 -> k2 -> *

type family RunX (a :: k1 -> k2 -> *) (b :: k1) :: k2
type instance RunX XId a = a
type instance RunX (XConst a) _ = a
type instance RunX (XMagic f) a = (RunX f) a a
type instance RunX (XCompose f1 f2) a = RunX f1 (RunX f2 a)

newtype XData t (a :: *) = XData {unXData :: RunX t a}

myFunction :: Rec (XData (XCompose (XMagic (XConst Maybe)) f)) ts -> Rec (XData f) ts -> Rec (XData f) ts
myFunction = rzipWith (\n o -> XData $ fromMaybe (unXData o) (unXData n))

test :: Rec (XData XId) [String, Int]
test = myFunction (XData (Just "bye") :& XData Nothing :& RNil) (XData "hi" :& XData 4 :& RNil)
```

**Comment by 2426021684** [comment]
@acowley Wouldn't it make sense to also handle `Const` in HKD? Why make the HKD closed? The fact that `Maybe a` is treated as `Maybe a` while `Identity a` is treated as `a` seems wierd. I feel that `Maybe a` should have to be escaped.

**Comment by acowley** [comment]
Good questions, all.

1. I didn't handle `Const` because I feel like it would be confusing. The main ones I wanted to handle are `Identity` and `Compose` because they are nothing but noise. I should perhaps not have included `ElField` since it is already handled pretty nicely. The reason I kept `ElField` is that if you have a type annotation, then this lets you avoid repeating the field names. For `Const`, I felt like seeing `("23", "A bool")` and getting a `Rec (Const String) '[Int, Bool]` was potentially confusing since it looks just like a `Rec Identity '[String, String]`. I can be convinced that this fear is misplaced, but I'd rather ease into this feature to see if it's already hiding too much, and then add more if it proves itself usable.

2. The family is closed so that we can have the `HasIsoHKDSym ` closed family with a catch-all case, and then use the result of that to write non-overlapping instances of `IsoHKD`. I think that if it was open, you would need to write a `type family instance HasIsoHKDSym MyFunctor = ...` for every context you want to use. Maybe I'm wrong about that, or maybe there's a better way to write it to provide a catch-all case so the user isn't forced to deal with this mechanism but still support extensibility.

3. How could mapping `Maybe a` to `a` work if you wanted to provide a value of `Nothing`?

My guiding principle was that `Identity` and `Compose` are a pain to use because they only serve as syntactic placeholders, so I wanted to get rid of that noise but not go so far as to obfuscate other things.

That said, I agree that closing the mechanism feels restrictive. I just didn't figure out how to make it trivial to use (i.e. the user doesn't *have* to do anything to use their type constructor as a context in a `Rec`) and allow users to participate in the `HKD` part.

**Comment by 2426021684** [comment]
@acowley What I'm saying is there should not be a catch all case. That would make its behavior clearer and allow it to be open.

**Comment by Lysxia** [comment]
With an open type family the catch-all case is replaced with a constructor that lifts type constructors to symbols for `HKD`:

    data TyCon f
    type instance HKD (TyCon f) a = f a

Then the original vinyl record would be equivalent to

    type Rec f = XRec (TyCon f)

Also note that `HKD` is [`Apply` from *singletons*](https://github.com/goldfirere/singletons/blob/23aa4bdaf05ce025a2493b35ec3c26cc94e3fdce/src/Data/Singletons/Internal.hs#L178).

**Comment by acowley** [comment]
You’re right, so you can define an `XMaybe` to make your record the shorter, but you don’t have to. 

@Lysxia It’s pretty similar to the way vinyl worked many years ago, pre-`singletons`! I’m not married to the name `HKD`, but chose it because I felt like it has some mindshare at the moment. We may as well change it to something Vinyl-specific to avoid clashes in case someone does import it into their code.

**Comment by acowley** [comment]
@2426021684 I really hope I didn't step on your toes by pushing back on the idea you outlined here; I truly am happy to include it as is. I think what we've got already is a huge improvement to vinyl's ergonomics. Moreover, the limitations are interesting! I don't know if you had a chance to look at it, but `rmap` is a curious case:

```haskell
rmap
  :: (forall x. f x -> g x)
  -> Rec f rs
  -> Rec g rs
rmap _ RNil = RNil
rmap η (x :& xs) = η x :& (η `rmap` xs)
```

The function must now contend with values captured in the records that are of the form, `RunX f x`. So we'd have `forall x. RunX f x -> RunX g x`, but `RunX` is not a very regular thing. The type of `rmap` says that whatever the actual type index of a record field was, it should be unchanged (e.g. we might turn `Identity Int -> Const String Int`, but the type index, `Int`, is the same). But do we know that `RunX` will do that? We can't, since `RunX` looks at the index (in order to return `a` from `Identity a`, for example), and is by no means injective. This means that the application `η x` in the cons case could return anything. I think we could concoct constraints on the behavior of `RunX g a` to nail this down, but I haven't done it yet. I think as far as I got was convincing myself that it would require QuantifiedConstraints which are Coming Soon (we want to say something like, `forall x. IndexOf (RunX g x) ~ x`).

Where Haskell is letting us down is in not simplifying type applications like `Id a ==> a` and compositions. I'm not sure that such computations are 100% incompatible with Haskell, so perhaps we'll get there.

I hope the compromise solution isn't too distasteful to you as I think it's terrific, and you're the reason it exists.

**Comment by 2426021684** [comment]
@acowley rmap works for me

```
f :: Rec (XData f) ts -> Rec (XData (XCompose (XMagic (XConst Maybe)) f)) ts
f = rmap (XData . Just . unXData)

t4 :: Rec (XData (XCompose (XMagic (XConst Maybe)) XId)) [String, Integer]
t4 = f (XData "hi" :& XData 5 :& RNil)
```

**Comment by acowley** [comment]
@2426021684 I meant with the approach where `XData` is baked into the record constructor so we could get the totally uncluttered Idris syntax. If we're willing to keep the uses of `XData` on types and values, then I don't think there's a problem, but it didn't strike me as particularly close to the desired Idris version.

**Comment by 2426021684** [comment]
@acowley That makes sense

**Comment by acowley** [comment]
@2426021684 Just to reiterate, I would be happy to include a version of this approach in vinyl if you would like to prepare a PR. I think it amounts to a separate module that offers `XData`, `RunX`, and the examples you've used in this thread. Giving a name to the composition of `XMagic . XConst` like `XLift` or `XTy` to riff on @Lysxia's suggestion seems like it might help intuitions.

**Comment by acowley** [comment]
@2426021684 I think you were right and I was wrong. I believe I've managed to do things more in the way you suggested (open type family!) without sacrificing any of the call-cite concision I fell in love with. That work is over on [this branch](https://github.com/VinylRecords/Vinyl/blob/srec-merge/Data/Vinyl/XRec.hs).

**Comment by 2426021684** [comment]
This seems interesting. http://h2.jaguarpaw.co.uk/posts/hkd-pattern-type-level-ski/

**Comment by acowley** [comment]
Any thoughts on the branch I linked with the `xrec` functions? I hope to push a release with that soon, and I’d be happy if you were happy with the usability.
